controllers/virtualmachine/virtualmachine/virtualmachine_controller.go

```diff
@@ -398,8 +401,16 @@ func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineContext) (reterr
 	// Upgrade schema fields where needed
 	upgradeSchema(ctx)
 
-	if err := r.VMProvider.CreateOrUpdateVirtualMachine(ctx, ctx.VM); err != nil {
-		r.Recorder.EmitEvent(ctx.VM, "CreateOrUpdate", err, false)
+	err := r.VMProvider.CreateOrUpdateVirtualMachine(ctx, ctx.VM)
+
+	if ctxop.IsCreate(ctx) {
+		r.Recorder.EmitEvent(ctx.VM, "Create", err, false)
+	} else if ctxop.IsUpdate(ctx) {
+		r.Recorder.EmitEvent(ctx.VM, "Update", err, false)
+	}
+
+	if err != nil {
+		r.Recorder.EmitEvent(ctx.VM, "CreateOrUpdate", err, true)
```

Comment:
Hi @ericvmw,

No, it generates distinct failure events as well. I just did not remove the old, combined event generated on failure in case anyone is pointing to it currently.

Reply:
@ericvmw , you can even see the tests we have for the distinct failure events at https://github.com/vmware-tanzu/vm-operator/pull/666/files#diff-cc296002fd0e8c72c0d657d91a474ee820f017e548bcaffeae4549cf76e97778R179-R218

Reply:
Great thanks @akutz !